lib/handlers/allow.js

```diff
@@ -93,56 +102,69 @@ function fetchDocument (host, ldp, baseUri) {
+ *   the user being delegated for).
+ */
+function getUserId (webId, onBehalfOf, fetchDocument) {
+  let principal = onBehalfOf
+  if (!webId || !principal) {
+    return Promise.resolve(webId)
   }
+  // WebID Delegation header present, verify secretary and return the principal
+  let secretary = webId
+  debug('Verifying delegator (secretary agent) ', secretary,
+    ' representing ', principal)
 
-  var delegator = utils.debrack(onBehalfOf)
-  verifyDelegator(req.hostname, delegator, req.session.userId,
-    function (err, res) {
-      if (err) {
-        err.status = 500
-        return callback(err)
-      }
-
-      if (res) {
-        debug('Request User ID (delegation) :' + delegator)
-        return callback(null, delegator)
+  return verifyDelegator(secretary, principal, fetchDocument)
+    .then(verified => {
+      if (verified) {
+        debug('Request User ID (via delegation): ' + principal)
+        return principal
+      } else {
+        debug('Secretary agent not authorized: ' + secretary)
+        return secretary
       }
-      return callback(null, req.session.userId)
     })
 }
 
-function verifyDelegator (host, ldp, baseUri, delegator, delegatee, callback) {
-  fetchDocument(host, ldp, baseUri)(delegator, function (err, delegatorGraph) {
-    // TODO handle error
-    if (err) {
-      err.status = 500
-      return callback(err)
-    }
-
-    var delegatesStatements = delegatorGraph
-      .each(delegatorGraph.sym(delegator),
-        delegatorGraph.sym('http://www.w3.org/ns/auth/acl#delegates'),
-        undefined)
-
-    for (var delegateeIndex in delegatesStatements) {
-      var delegateeValue = delegatesStatements[delegateeIndex]
-      if (utils.debrack(delegateeValue.toString()) === delegatee) {
-        callback(null, true)
+/**
+ * @see https://www.w3.org/wiki/WebID/Delegation
+ * @see https://github.com/seebi/WebID-Delegation-paper
+ * @param secretary {string} Web ID of the delegator agent. Acts on behalf of the
+ *   principal.
+ * @param principal {string} Web ID of the secretary's "owner". Authorizes
+ *   the secretary to act on the principal's behalf.
+ * @param fetchDocument {Function} Result of `documentFetcher()` call
+ * @throws {Error}
+ * @return {Promise<Boolean>} Whether or not delegator is verified
+ */
+function verifyDelegator (secretary, principal, fetchDocument) {
+  return new Promise((resolve, reject) => {
+    // Fetch the principal's WebID Profile document
+    fetchDocument(principal, (err, profileGraph) => {
+      if (err) {
+        debug("Error fetching the principal's profile: ", err)
+        err.status = 500
+        return reject(err)
       }
-    }
-    // TODO check if this should be false
-    return callback(null, false)
+      let authorizedMatches = profileGraph
+        .match(
+          rdf.namedNode(principal),
+          ns.acl('delegates'),
+          rdf.namedNode(secretary)
+        )
+      let isAuthorized = authorizedMatches.length > 0
```

Comment:
Great. I think you could squish these two statements into one using the `.any()`.